### PR TITLE
Implement appender support

### PR DESCRIPTION
**Pull request by JinnyYi**
Implement appender support in ref: https://github.com/aos-dev/go-storage/issues/529.

### CODE REVIEW

**Comment by Xuanwo** [comment]
Please take https://github.com/aos-dev/go-service-fs/blob/master/tests/storage_test.go#L17-L22 a look, run an integration test for oss.

File: storage.go [hunk 1]
```diff
@@ -22,6 +22,35 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
 	return o
 }
 
+func (s *Storage) createAppend(ctx context.Context, path string, opt pairStorageCreateAppend) (o *Object, err error) {
+	rp := s.getAbsPath(path)
+
+	var nextPos int64 = 0
+	isExist, err := s.bucket.IsObjectExist(rp)
```
Comment (Xuanwo) [inline storage.go:29]: We don't need to check if this object exists or not, just return error that we got.

File: storage.go [hunk 2]
```diff
@@ -22,6 +22,35 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
...
+	if err != nil {
+		return
+	}
+
+	if isExist {
+		props, errGetMeta := s.bucket.GetObjectDetailedMeta(rp)
```
Comment (Xuanwo) [inline storage.go:35]: In `createAppend`, we can use `AppendObject` with position 0 and size 0 to create an Object.

File: storage.go [hunk 3]
```diff
@@ -216,3 +245,20 @@ func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int6
 	}
 	return size, nil
 }
+
+func (s *Storage) writeAppend(ctx context.Context, o *Object, r io.Reader, size int64, opt pairStorageWriteAppend) (n int64, err error) {
+	rp := o.GetID()
+	nextPos := o.MustGetAppendOffset()
```
Comment (Xuanwo) [inline storage.go:251]: `MustGetAppendOffset` will panic if append offset does not exist.

In production code, we usually `offset, ok := o.GetAppendOffset()` and check the returning value to decide whether to return an error or not.

File: storage.go (hunk 2 above)
Comment (JinnyYi) [reply storage.go:35]: Does go-service-oss need to support append to an existing object?

Comment (Xuanwo) [reply storage.go:35]: Yes, but not in this way.

`CreateAppend` needs to `create` an append object and should not do other jobs.

In order to append to an existing object, use can:

- Use `Stat` to get an append object.
- Use `Create` to create an append object if they know the offset.

File: storage.go (hunk 3 above)
Comment (JinnyYi) [reply storage.go:251]: Which error is suitable when the append offset is unset? Do I need to definite error code?

Comment (Xuanwo) [reply storage.go:251]: For now, let's just use `fmt.Errorf("append offset is not set")`. We will add this error in `go-storage`.

File: Makefile [hunk 4]
```diff
@@ -32,7 +32,7 @@ test:
 	go tool cover -html="coverage.txt" -o "coverage.html"
 
 integration_test:
-	go test -count=1 -race -covermode=atomic -v ./tests
+	STORAGE_OSS_INTEGRATION_TEST=on go test -count=1 -race -covermode=atomic -v ./tests
```
Comment (Xuanwo) [inline Makefile:35]: We don't need to set `STORAGE_OSS_INTEGRATION_TEST=on` here, we will use `Makefile.env`

File: storage.go [hunk 5]
```diff
@@ -22,6 +23,25 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
 	return o
 }
 
+func (s *Storage) createAppend(ctx context.Context, path string, opt pairStorageCreateAppend) (o *Object, err error) {
+	rp := s.getAbsPath(path)
+
+	options := make([]oss.Option, 0)
+	options = append(options, oss.ContentLength(0))
+
+	offset, err := s.bucket.AppendObject(rp, strings.NewReader(""), 0, options...)
```
Comment (Xuanwo) [inline storage.go:32]: Can we use `nil` instead of `strings.NewReader("")` here?

Comment (JinnyYi) [reply storage.go:32]: Maybe the GetReaderLen() in https://github.com/aliyun/aliyun-oss-go-sdk/blob/master/oss/utils.go#L438 will return error if we use nil?

Comment (Xuanwo) [reply storage.go:32]: Let's give it a try. Our integration tests will cover it.

File: storage.go [hunk 6]
```diff
@@ -24,21 +23,11 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
...
-	options := make([]oss.Option, 0)
-	options = append(options, oss.ContentLength(0))
-
-	offset, err := s.bucket.AppendObject(rp, strings.NewReader(""), 0, options...)
-	if err != nil {
-		return
-	}
-
-	o = s.newObject(true)
+	o = s.newObject(false)
```
Comment (Xuanwo) [inline storage.go:26]: We should create this object with `true`, as there is no more info we cloud provide.

**Comment by Xuanwo** [comment]
Other LGTM, please don't forget to update the docs.

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**